tachyon/crypto/hashes/sponge/poseidon/poseidon_config.h

```diff
@@ -62,13 +69,23 @@ struct PoseidonConfig : public PoseidonConfigBase<F> {
   // Maximally Distance Separating (MDS) Matrix.
   math::Matrix<F> mds;
 
-  PoseidonConfig() = default;
-  PoseidonConfig(const PoseidonConfigBase<F>& base, const math::Matrix<F>& mds)
-      : PoseidonConfigBase<F>(base), mds(mds) {}
-  PoseidonConfig(PoseidonConfigBase<F>&& base, math::Matrix<F>&& mds)
-      : PoseidonConfigBase<F>(std::move(base)), mds(std::move(mds)) {}
+  // See
+  // https://github.com/privacy-scaling-explorations/poseidon/blob/d29f35d9744e0c58b63679a15730ff2fab470ef2/src/spec.rs#L71.
+  math::Matrix<F> optimized_ark;
 
-  static PoseidonConfig CreateDefault(size_t rate, bool optimized_for_weights) {
+  PoseidonConfig() = default;
+  PoseidonConfig(const PoseidonConfigBase<F>& base, const math::Matrix<F>& mds,
+                 const math::Matrix<F>& optimized_ark)
+      : PoseidonConfigBase<F>(base), mds(mds), optimized_ark(optimized_ark) {}
+  PoseidonConfig(PoseidonConfigBase<F>&& base, math::Matrix<F>&& mds,
+                 math::Matrix<F>&& optimized_ark)
+      : PoseidonConfigBase<F>(std::move(base)),
+        mds(std::move(mds)),
+        optimized_ark(std::move(optimized_ark)) {}
+
+  static PoseidonConfig CreateDefault(
+      size_t rate, bool optimized_for_weights,
+      bool use_efficient_implementation = false) {
```

Comment:
Why would anyone not use efficient implementation?

Reply:
Good question! and I applied this efficient implementation to all!